File: gogame.cpp
#include <cstdio>
#include <string.h>
#include <stdlib.h>
#include <assert.h> //FIXME

#include "gogame.h"

Vertex::Vertex()
{
	x = -1;
	y = -1;
}

Vertex::Vertex(int _x, int _y)
{
	x = _x;
	y = _y;
}

int distanceFromEdge(Vertex v)
{
	int dx = HALF_INDEX-abs(HALF_INDEX-v.x);
	int dy = HALF_INDEX-abs(HALF_INDEX-v.y);
	
	return min(dx,dy);
}

int v2i(Vertex v)
{
	return v.y*BOARD_SIZE + v.x;
}

Move::Move()
{
	p = EMPTY;
}

Move::Move(Player _p, Vertex _v)
{
	p = _p;
	v = _v;
}

void Move::print()
{
	printf("%s: %02d %02d\n", p==B? "B" : "W", v.x, v.y);
}

GoGame::GoGame(int _handicap, int _numMoves, Move *_moves)
{
	handicap = _handicap;
	numMoves = _numMoves;
	moves = (Move*)malloc(numMoves*sizeof(Move));
	memcpy(moves, _moves, numMoves*sizeof(Move));
	
	states = (Board**)malloc(numMoves*sizeof(Board*));
	states[0] = new Board(); //FIXME assuming B first, no handicap
	for(int i=1; i<numMoves; i++) {
		//moves[i-1].print();
		states[i] = new Board(states[i-1], moves[i-1]);
	}
	
	/*for(int i=0; i<numMoves; i++) {
		moves[i].print();
		//states[i]->print();
	}*/
}

GoGame::~GoGame()
{
	free(moves);
	for(int i=0; i<numMoves; i++) {
		delete states[i];
	}
	free(states);
}


Board::Board()
{
	grid = (char*)calloc(BOARD_SIZE*BOARD_SIZE, sizeof(char));
	nextPlayer = B;
}

Board::Board(Board *prevBoard, Move m)
{
	int numBytes = BOARD_SIZE*BOARD_SIZE*sizeof(char);
	grid = (char*)malloc(numBytes);
	memcpy(grid, prevBoard->grid, numBytes);
	nextPlayer = FLIP_PLAYER(prevBoard->nextPlayer);
	playMove(m);
}

Board::~Board()
{
	free(grid);
}

void Board::playMove(Move m)
{
	//NOTE ASSUME M IS LEGAL!
	Vertex v = m.v;
	grid[v2i(v)] = m.p;
	
	//check if life status of adjacent enemy groups changed
	Vertex v2;
	bool visited[BOARD_SIZE*BOARD_SIZE];
	int numCapped = 0;
	if(getW(v, v2)) {
		if(grid[v2i(v2)] == nextPlayer) {
			bzero(visited, BOARD_SIZE*BOARD_SIZE);
			numCapped += removeIfCaptured(v2, visited);
		}
	}
	if(getE(v, v2)) {
		if(grid[v2i(v2)] == nextPlayer) {
			bzero(visited, BOARD_SIZE*BOARD_SIZE);
			numCapped += removeIfCaptured(v2, visited);
		}
	}
	if(getN(v, v2)) {
		if(grid[v2i(v2)] == nextPlayer) {
			bzero(visited, BOARD_SIZE*BOARD_SIZE);
			numCapped += removeIfCaptured(v2, visited);
		}
	}
	if(getS(v, v2)) {
		if(grid[v2i(v2)] == nextPlayer) {
			bzero(visited, BOARD_SIZE*BOARD_SIZE);
			numCapped += removeIfCaptured(v2, visited);
		}
	}
	
	/*if(numCapped > 0) {
		printf("numCapped: %d\n", numCapped);
	}*/
}


int Board::removeIfCaptured(Vertex v, bool * visited)
{
	//Assume v is on board, belongs to enemy
	list<Vertex> stoneList;
	
	if(findLiberty(v, visited, &stoneList) == true) {
		return 0;
	} else {
		for (list<Vertex>::iterator it = stoneList.begin(); it != stoneList.end(); it++) {
    		//printf("kill: %d %d\n", (*it).x, (*it).y);
    		grid[v2i(*it)] = EMPTY;
    	}
		
		return stoneList.size(); //how many were captured
	}
}


bool Board::findLiberty(Vertex v, bool * visited, list<Vertex>* stoneList)
{
	visited[v2i(v)] = true;
	
	Vertex v2;
	int i2;
	
	if(getW(v, v2)) {
		i2 = v2i(v2);
		if(grid[i2] == EMPTY) {
			return true;
		} else if(grid[i2] == nextPlayer && !visited[i2]) {
			if(findLiberty(v2, visited, stoneList)) {
				return true;
			}
		}
	}
	if(getE(v, v2)) {
		i2 = v2i(v2);
		if(grid[i2] == EMPTY) {
			return true;
		} else if(grid[i2] == nextPlayer && !visited[i2]) {
			if(findLiberty(v2, visited, stoneList)) {
				return true;
			}
		}
	}
	if(getN(v, v2)) {
		i2 = v2i(v2);
		if(grid[i2] == EMPTY) {
			return true;
		} else if(grid[i2] == nextPlayer && !visited[i2]) {
			if(findLiberty(v2, visited, stoneList)) {
				return true;
			}
		}
	}
	if(getS(v, v2)) {
		i2 = v2i(v2);
		if(grid[i2] == EMPTY) {
			return true;
		} else if(grid[i2] == nextPlayer && !visited[i2]) {
			if(findLiberty(v2, visited, stoneList)) {
				return true;
			}
		}
	}
	
	stoneList->push_back(v);
	return false;
}


void Board::print()
{
	char * g = grid;
	for(int y=0; y<BOARD_SIZE; y++) {
		for(int x=0; x<BOARD_SIZE; x++) {
			switch(*g) {
				case 0:
					printf(".");
					break;
				case B:
					printf("X");
					break;
				case W:
					printf("O");
					break;
				default:
					printf("?");
					
			}
			g++;
		}
		printf("\n");
	}
}


inline bool Board::getW(Vertex &v1, Vertex &v2)
{
	if(v1.x <= 0) {
		return false;
	} else {
		v2.x = v1.x-1;
		v2.y = v1.y;
		return true;
	}
}

inline bool Board::getE(Vertex &v1, Vertex &v2)
{
	if(v1.x >= (BOARD_SIZE-1)) {
		return false;
	} else {
		v2.x = v1.x+1;
		v2.y = v1.y;
		return true;
	}
}

inline bool Board::getN(Vertex &v1, Vertex &v2)
{
	if(v1.y <= 0) {
		return false;
	} else {
		v2.x = v1.x;
		v2.y = v1.y-1;
		return true;
	}
}

inline bool Board::getS(Vertex &v1, Vertex &v2)
{
	if(v1.y >= (BOARD_SIZE-1)) {
		return false;
	} else {
		v2.x = v1.x;
		v2.y = v1.y+1;
		return true;
	}
}

File: gogame.py
from os import path
from sgf import SGF
import numpy as np
import logging
import time
from govea import B, W, opposite_color


GRID_DTYPE = np.int8


class GoGame(object):
	def __init__(self, sgf, debug=False):
		#iterate through sgf.moves, creating self.boardstates
		#first boardstate is empty, or has handicap stones
		#move i gets applied to state i to produce state i+1
		#so if we have m moves, we have m+1 states
		self.board_shape = sgf.board_shape
		self.handicap = sgf.handicap
		self.komi = sgf.komi
		self.player_w = sgf.player_w
		self.player_b = sgf.player_b
		self.date = sgf.date
		self.result = sgf.result
		self.moves = sgf.moves
		#TODO copy more fields

		initial_state = BoardState.empty_board(self.board_shape)
		for m in sgf.initial_moves:
			initial_state.grid[m.r,m.c] = m.color

		self.states = [initial_state]
		prev_state = initial_state
		
		for i, m in enumerate(self.moves):
			if debug: #FIXME
				print "about to play move %d: %s" % (i, m)
			try:
				state = BoardState(prev_state, m)
			except InconsistentBoardStateError as e:
				e.msg = "Inconsistent board state noticed at move %d (%s) in sgf %s" % (i, m, sgf.filename)
				raise e
			if debug:
				print "resulting state:"
				print state
				time.sleep(1.0)
			self.states.append(state)
			prev_state = state


		#or s in self.states:
		#	print(s)


class BoardState(object):
	def __init__(self, prev_state, move, shape=None):
		#board state results from applying move to prev_state
		self.prev_move = move
		if prev_state is None and move is None:
			self.shape = shape
			self.grid = np.zeros(shape, dtype=GRID_DTYPE)
			self._create_neighbor_map(shape)
			self.num_played_moves = 0
		else:
			self.shape = prev_state.shape
			self.grid = prev_state.grid.copy()
			self.neighbor_map = prev_state.neighbor_map #no need to copy this one
			self.num_played_moves = prev_state.num_played_moves + 1
			#print self.num_played_moves, move
			if not move.is_pass():
				if self.grid[move.r, move.c] != 0:
					raise InconsistentBoardStateError("")
				color = move.color #the color of the player making the move
				enemy_color = opposite_color(color)
				r = move.r
				c = move.c
				self.grid[r,c] = color
				#assume move is legal. (not suicide, etc.)
				#for each adjacent enemy group, check to see if it has any liberties left
				#(send in a blank visit grid)
				#if not, remove that group and increment captures appropriately
				visited = np.zeros(self.shape, dtype=np.bool)
				for r2, c2 in self.neighbor_map[r,c]:
					if self.grid[r2,c2] == enemy_color:
						#it's an enemy group
						if visited[r2,c2]:
							continue #we've already thought about this group
						visited.fill(False)
						if not self._has_liberty(r2, c2, enemy_color, visited):
							#this enemy group has been killed!
							num_removed = np.count_nonzero(visited) #FIXME add to tally
							self.grid[visited] = 0
							logging.debug('Move %s killed %d stones', move, num_removed)
			
			#time.sleep(1.0)
			#print self
			

	@classmethod
	def empty_board(cls, shape):
		return cls(None, None, shape=shape)


	def _create_neighbor_map(self, shape):
		self.neighbor_map = np.ndarray(shape, dtype=list)
		for r in range(shape[0]):
			for c in range(shape[1]): #TODO shortcut?
				neighbors = [] #will have 2,3, or 4 neighbors
				
				if r > 0: #Up
					neighbors.append((r-1,c))
				if c > 0: #Left
					neighbors.append((r,c-1))
				if r < shape[0]-1: #Down
					neighbors.append((r+1,c))
				if c < shape[1]-1: #Right
					neighbors.append((r,c+1))

				self.neighbor_map[r,c] = neighbors


	def __repr__(self):
		#TODO how much overhead in this approach?
		np.set_printoptions(formatter={'int_kind':lambda x: {B:'X', W:'O', 0:'.'}[x]})
		s = str(self.grid)
		np.set_printoptions() #reset to default
		return s



	def _has_liberty(self, r, c, color, visited):
		#return true if at least one liberty exists for color's group at r,c
		#modifies visited grid in the process
		if visited[r,c]:
			return False #we've already been here
		visited[r,c] = True
		#logging.debug("visiting %d %d (%d)", r, c, self.grid[r,c])
		
		for r2, c2 in self.neighbor_map[r,c]:
			x = self.grid[r2,c2]
			#logging.debug("considering %d %d (%d)", r2, c2, x)
			if x == 0 or (x==color and self._has_liberty(r2, c2, color, visited)):
				return True
		
		return False #exhausted all possibilities, give up

		

class InconsistentBoardStateError(Exception):
	def __init__(self, msg):
		self.msg = msg

	def __str__(self):
		return repr(self.msg)


if __name__ == '__main__':
	with open(path.join('data', 'sgf', 'Hutoshi4-kghin.sgf')) as f:
		sgf = SGF(f)
		game = GoGame(sgf)

